File: lang/go/cxlib/id_abi_test.go
package cxlib

import (
	"encoding/json"
	"strings"
	"testing"
)

// Tests for the ID/IDREF C ABI surface (cx_id_lookup / cx_resolve_ref /
// cx_node_id) / Phase 7.65. Document-level resolve is
// already tested by identity_test.go; these exercise the stateless
// string-in/string-out C ABI variants.

const idAbiDoc = `[users
  [user #u-1 name=alice]
  [user #u-2 name=bob]
  [reviewer assigned-to=@u-1]
]`

func TestIDLookupReturnsAstJSON(t *testing.T) {
	out, err := IDLookup(idAbiDoc, "u-1")
	if err != nil {
		t.Fatal(err)
	}
	if out == "" {
		t.Fatal("expected non-empty result")
	}
	var payload map[string]interface{}
	if err := json.Unmarshal([]byte(out), &payload); err != nil {
		t.Fatalf("invalid JSON: %v", err)
	}
	if payload["type"] != "Element" || payload["name"] != "user" || payload["id"] != "u-1" {
		t.Fatalf("payload = %v", payload)
	}
}

func TestIDLookupMissingReturnsEmpty(t *testing.T) {
	out, err := IDLookup(idAbiDoc, "does-not-exist")
	if err != nil {
		t.Fatal(err)
	}
	if out != "" {
		t.Fatalf("expected empty, got %q", out)
	}
}

func TestResolveRefMatchesIDLookup(t *testing.T) {
	a, err := IDLookup(idAbiDoc, "u-2")
	if err != nil {
		t.Fatal(err)
	}
	b, err := ResolveRef(idAbiDoc, "u-2")
	if err != nil {
		t.Fatal(err)
	}
	if a != b {
		t.Fatalf("resolve_ref %q != id_lookup %q", b, a)
	}
}

// TestNodeID* removed — cx_node_id was retired
// alongside the cxpath C ABI. Equivalent: EvalCode with a CXPath
// `//pattern` value or a `[?for [pattern $m] :yield $m]`
// comprehension, then read $m/@id.

func TestIDLookupParseError(t *testing.T) {
	_, err := IDLookup("[unclosed", "u-1")
	if err == nil {
		t.Fatal("expected parse error")
	}
	if !strings.Contains(err.Error(), "expect") && !strings.Contains(err.Error(), "parse") && !strings.Contains(err.Error(), "unexpected") && !strings.Contains(err.Error(), "EOF") {
		t.Logf("error: %v", err)
	}
}

File: lang/rust/cxlib/tests/id_abi_test.rs
//! Integration tests for the Phase 7.65 ID/IDREF C ABI wrappers
//! (`cx_id_lookup`, `cx_resolve_ref`, `cx_node_id`) per ADR 0003.

use cxlib::{id_lookup, resolve_ref, node_id};
use serde_json::Value;

const DOC: &str = "[users\n  [user #u-1 name=alice]\n  [user #u-2 name=bob]\n  [reviewer assigned-to=@u-1]\n]\n";

#[test]
fn test_id_lookup_happy_path() {
    let out = id_lookup(DOC, "u-1").expect("id_lookup ok");
    let s = out.expect("id_lookup found u-1");
    let v: Value = serde_json::from_str(&s).expect("AST-JSON parses");
    assert_eq!(v["type"], "Element");
    assert_eq!(v["name"], "user");
    assert_eq!(v["id"], "u-1");
}

#[test]
fn test_id_lookup_missing_returns_none() {
    let out = id_lookup(DOC, "does-not-exist").expect("id_lookup ok");
    assert!(out.is_none(), "missing id should return Ok(None), got {:?}", out);
}

#[test]
fn test_resolve_ref_equals_id_lookup() {
    let a = id_lookup(DOC, "u-2").expect("id_lookup ok").expect("u-2 found");
    let b = resolve_ref(DOC, "u-2").expect("resolve_ref ok").expect("u-2 found");
    assert_eq!(a, b, "resolve_ref must equal id_lookup for shared namespace");
}

#[test]
fn test_node_id_at_cxpath() {
    let user = node_id(DOC, "//user").expect("node_id ok");
    assert_eq!(user.as_deref(), Some("u-1"));

    let reviewer = node_id(DOC, "//reviewer").expect("node_id ok");
    assert!(reviewer.is_none(), "reviewer has no id, expected None, got {:?}", reviewer);
}
